apps/server/src/modules/tool/uc/external-tool.uc.ts

```diff
@@ -0,0 +1,67 @@
+import { Inject, Injectable, UnprocessableEntityException } from '@nestjs/common';
+import { ICurrentUser, Permission, User } from '@shared/domain';
+import { ExternalToolService } from '@src/modules/tool/service/external-tool.service';
+import { ExternalToolDO, Lti11ToolConfigDO, Oauth2ToolConfigDO } from '@shared/domain/domainobject/external-tool';
+import { AuthorizationService } from '@src/modules/authorization';
+import { OauthProviderService } from '@shared/infra/oauth-provider';
+import { ProviderOauthClient } from '@shared/infra/oauth-provider/dto';
+import { DefaultEncryptionService, IEncryptionService } from '@shared/infra/encryption';
+import { ExternalToolRequestMapper } from '@src/modules/tool/mapper/external-tool-request.mapper';
+
+@Injectable()
+export class ExternalToolUc {
+	constructor(
+		private readonly externalToolService: ExternalToolService,
+		private readonly externalToolMapper: ExternalToolRequestMapper,
+		private readonly authorizationService: AuthorizationService,
+		private readonly oauthProviderService: OauthProviderService,
+		@Inject(DefaultEncryptionService) private readonly oAuthEncryptionService: IEncryptionService
+	) {}
+
+	async createExternalTool(externalToolDO: ExternalToolDO, currentUser: ICurrentUser): Promise<ExternalToolDO> {
+		const user: User = await this.authorizationService.getUserWithPermissions(currentUser.userId);
+		this.authorizationService.checkAllPermissions(user, [Permission.TOOL_CREATE]);
+
+		if (!(await this.externalToolService.isNameUnique(externalToolDO))) {
```

Comment:
I would like to see all validations in a private function :)

Reply:
done